### PR TITLE
Alerts

### PR DESCRIPTION
**Pull request by dslowikowski**

### CODE REVIEW

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by twymer** [comment]
I did a pass through this today, didn't really see anything that stood out a lot, but had some trouble figuring out what was going on and sort of trying to understand the pretty repetitive code. Going to give it another look through in the morning.

**Comment by czue** [comment]
@twymer did you ever take another pass? i'm also happy to do one if helpful.

File: custom/ewsghana/alerts/alerts.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+    now = datetime.datetime.utcnow()
+    reported_products = StockTransaction.objects.filter(case_id=sp_id, type='stockonhand', report__date=now)
+    expected_products = SQLProduct.objects.filter(domain=user.domain)
+    reported_products_ = set()
+    expected_products_ = set()
+    for product in reported_products:
+        reported_products_.add(SQLProduct.objects.get(product_id=product.product_id).name)
+    for product in expected_products:
+        expected_products_.add(SQLProduct.objects.get(product_id=product.product_id).name)
+    missing_products = set.difference(reported_products_, expected_products_)
```
Comment (twymer) [inline custom/ewsghana/alerts/alerts.py]: This method is pretty confusing and going to cause an enormous number of unnecessary queries.

`reported_products` is a list of transactions
`expected_products` is a list of `SQLProduct`s that you then iterate over to refetch all `SQLProduct`s in that list and add their names to a set..

You should just get the `expected_products` set by: 
`set(SQLProduct.objects.filter(domain=user.domain).values_list('name', flat=True))` 
which will get the list that you're getting, but only the names, and as a flat list, then convert that whole thing to a set.

Getting the set for `reported_products` is a bit trickier, but you can do basically the same:
`set(StockTransaction.objects.filter(case_id=sp_id, type='stockonhand', report__date=now).values_list('sql_product__name', flat=True))`

File: custom/ewsghana/alerts/alerts.py [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+                        sp = facility.supply_point_id
+                        if sp:
+                            stocked_out = StockTransaction.objects.filter(case_id=sp, type="stockonhand",
+                                                                          stock_on_hand=0)
+                            if stocked_out.exists():
+                                no_rep += 1
+                                fac.add(str(facility))
+                            for product in stocked_out:
+                                stocked_out_products.add(SQLProduct.objects.get(
+                                    product_id=product.product_id).name)
```
Comment (twymer) [inline custom/ewsghana/alerts/alerts.py]: See my other comment about making set objects like this.

File: custom/ewsghana/alerts/alerts.py [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
```
Comment (twymer) [inline custom/ewsghana/alerts/alerts.py]: See my other comment about making set objects like this.

File: custom/ewsghana/alerts/alerts.py [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+                        sp = facility.supply_point_id
+                        if sp:
+                            stocked_out = StockTransaction.objects.filter(
+                                case_id=sp, type="stockonhand", stock_on_hand=0)
+                            if stocked_out.exists():
+                                no_rep += 1
+                                fac.add(str(facility))
+                                for product in stocked_out:
+                                    stocked_out_products.add(
+                                        SQLProduct.objects.get(product_id=product.product_id).name)
```
Comment (twymer) [inline custom/ewsghana/alerts/alerts.py]: See my other comment about making set objects like this.

**Comment by twymer** [comment]
@dslowikowski Did you see my comments?

In case @dslowikowski is offline, @lwyszomi / @agrzywinski

**Comment by agrzywinski** [comment]
@twymer Darek is offline for a while. I will lok on it tomorrow.

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by agrzywinski** [comment]
@twymer I've updated PR.

**Pull request merged**